src/controllers/reviewController.js

```diff
@@ -0,0 +1,119 @@
+    user: req.user.userId,
+  });
+
+  if (alreadySubmitted) {
+    throw new CustomError.Conflict(
+      'You already submitted review for this album'
+    );
+  }
+  //The user property is added to the request body with the userId of the authenticated user. This ensures that the created review is associated with the correct user.
+  req.body.user = req.user.userId; // req.user.userId came from req.user = { name, userId, role }; //from the object we pass with each request to the browser with name, userId role, - check authentication.js line 16 // so we are able to see in postman which user makes this post request- "user": "123"
+  const review = await Review.create(req.body);
+  res.status(StatusCodes.CREATED).json({ review });
+};
+
+//add get all reviews
+
+//update review
+const updateReview = async (req, res) => {
+  const { id: reviewId } = req.params; //take id of the review from req.params and assign it to reviewId
+
+  const review = await Review.findOne({ _id: reviewId });
+
+  if (!review) {
+    throw new CustomError.NotFoundError(`No review with id ${reviewId}`);
+  }
+
+  // Check if the requesting user is the author of the review
+  //option 1
+  if (review.user.toString() !== req.user.userId) {
+    throw new CustomError.UnauthorizedError(
+      'You are not authorized to update this review'
+    );
+  }
+  //option 2
+  // checkPermissions(req.user, review.user);
+
+  // Validate if required fields are present in the request body
+  //option A:
+  if (!req.body.rating || !req.body.title || !req.body.comment) {
+    throw new CustomError.BadRequestError(
+      'Rating, title, and comment are required fields'
+    );
+  }
+  // Update review properties
+  review.rating = req.body.rating;
+  review.title = req.body.title;
+  review.comment = req.body.comment;
+
+  //option B:
+  //   review.rating = req.body.hasOwnProperty('rating')
+  //     ? req.body.rating
+  //     : review.rating;
+  //   review.title = req.body.hasOwnProperty('title')
+  //     ? req.body.title
+  //     : review.title;
+  //   review.comment = req.body.hasOwnProperty('comment')
+  //     ? req.body.comment
+  //     : review.comment;
+
+  await review.save();
+  res.status(StatusCodes.OK).json({ review });
+};
+
+// get All Reviews
+const getAllReviews = async (req, res) => {
+  const reviews = await Review.find({}).populate({
+    //adding more info to the review using populate method- aka fill , add the info from mongoose model
+   //review+ about which album 
+   path: 'album', //we pass what we want to reference //line 26 in Review.js
+   select: 'artistName albumName image price', //and what properties we want to get from Album model (Album.js)
+
+    //and to populate user data: review+user who wrote it
+    // path: 'user', //we pass what we want to reference //in Review.js
+    // select: 'email name username', //and what properties we want to get from album model
+  });
+  res.status(StatusCodes.OK).json({ reviews, count: reviews.length });
+};
```

Comment:
We're fine to leave this endpoint+controller, since I can imagine there can be cases where the frontend want sto get _all_ the reviews. But for [PS-73,](https://codethedreamteam4-2023.atlassian.net/browse/PS-73) we specifically want an endpoint that gives the client just the reviews for a particular product/album. That way we reduce the size of the response we're returning and the client doesn't have to filter out the results list themselves. Filtering at the db-level is generally faster than filtering in client-side javascript code.

Reply:
@akosasante should I add review property yo Album.js and also add virtuals to retrieve reviews for a single album?

Reply:
That would work I think. I believe a simpler approach for now would just be to use the `Review` model since it already has an `album` property. So we can do something like `Review.find({ album: albumId })` to fetch all reviews for a particular album and return those review objects. 

Your approach does have the benefit of then the client only has to make a single request on the album details page for that information.

Reply:
@akosasante did it! just commited to GitHub, so I think this was https://codethedreamteam4-2023.atlassian.net/browse/PS-76